### PR TITLE
Static pod creation

### PR DESCRIPTION
**Pull request by bpickard22**
fixes https://issues.redhat.com/browse/OCPBUGS-5042

### CODE REVIEW

File: go-controller/pkg/cni/cni.go [hunk 1]
```diff
@@ -136,8 +136,11 @@ func (pr *PodRequest) cmdAdd(kubeAuth *KubeAPIAuth, podLister corev1listers.PodL
 	if err != nil {
 		return nil, fmt.Errorf("failed to get pod annotation: %v", err)
 	}
-	if err := pr.checkOrUpdatePodUID(podUID); err != nil {
-		return nil, err
+	pod, err := getPod(podLister, kclient, pr.PodNamespace, pr.PodName)
+	if !IsStaticPod(pod) && err == nil {
+		if err := pr.checkOrUpdatePodUID(podUID); err != nil {
```
Comment (flavio-fernandes) [inline go-controller/pkg/cni/cni.go:141]: NIT: remove the ':' from ':='

File: go-controller/pkg/cni/cni.go [hunk 2]
```diff
@@ -136,8 +136,11 @@ func (pr *PodRequest) cmdAdd(kubeAuth *KubeAPIAuth, podLister corev1listers.PodL
 	if err != nil {
 		return nil, fmt.Errorf("failed to get pod annotation: %v", err)
 	}
-	if err := pr.checkOrUpdatePodUID(podUID); err != nil {
-		return nil, err
+	pod, err := getPod(podLister, kclient, pr.PodNamespace, pr.PodName)
+	if !IsStaticPod(pod) && err == nil {
```
Comment (flavio-fernandes) [inline go-controller/pkg/cni/cni.go:140]: check if err == nil first:  `if err == nil && !IsStaticPod(pod) {`

**Review by flavio-fernandes — COMMENTED** [review]
initial comments

File: go-controller/pkg/cni/utils.go [hunk 3]
```diff
@@ -133,6 +133,7 @@ func PodAnnotation2PodInfo(podAnnotation map[string]string, checkExtIDs bool, po
 }
 
 //Taken from https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/types/pod_update.go
+//Start
```
Comment (flavio-fernandes) [inline go-controller/pkg/cni/utils.go:136]: move this to the previous commit:  `[Add Functions needed to check for static pod](https://github.com/ovn-org/ovn-kubernetes/pull/3347/commits/d94386ae0542a8a2d088d34aa5d231d09dd0c4fc)`

File: go-controller/pkg/cni/cni.go (hunk 1 above)
Comment (bpickard22) [reply go-controller/pkg/cni/cni.go:141]: im declaring and assigning the variable pod here so I need the := I think

File: go-controller/pkg/cni/utils.go [hunk 4]
```diff
@@ -131,3 +131,38 @@ func PodAnnotation2PodInfo(podAnnotation map[string]string, checkExtIDs bool, po
...
+		}
+	}
+	return "", fmt.Errorf("cannot get source of pod %q", pod.UID)
+}
+
+// IsStaticPod returns true if the pod is a static pod.
+func IsStaticPod(pod *kapi.Pod) bool {
+	source, err := GetPodSource(pod)
+	return err == nil && source != ApiserverSource
+}
```
Comment (flavio-fernandes) [inline go-controller/pkg/cni/utils.go:169]: `// END Taken from https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/types/pod_update.go`

File: go-controller/pkg/cni/utils.go [hunk 5]
```diff
@@ -131,3 +131,40 @@ func PodAnnotation2PodInfo(podAnnotation map[string]string, checkExtIDs bool, po
...
+	ConfigHashAnnotationKey      = "kubernetes.io/config.hash"
+)
+
+const (
+	// Filesource idenitified updates from a file.
+	FileSource = "file"
+	// HTTPSource identifies updates from querying a web page.
+	HTTPSource = "http"
+	// ApiserverSource identifies updates from Kubernetes API Server.
+	ApiserverSource = "api"
```
Comment (tssurya) [inline go-controller/pkg/cni/utils.go:149]: same here do you want all? only ApiserverSource right? and you can combine the two constants in the same block

**Review by tssurya — CHANGES_REQUESTED** [review]
Thanks Ben for the PR.
Let's get some e2e's and unit tests done for this. Surely we aren't having anything running currently else we would have seen conformance suite failures?

File: go-controller/pkg/cni/cni.go [hunk 6]
```diff
@@ -136,8 +136,11 @@ func (pr *PodRequest) cmdAdd(kubeAuth *KubeAPIAuth, podLister corev1listers.PodL
 	if err != nil {
 		return nil, fmt.Errorf("failed to get pod annotation: %v", err)
 	}
-	if err := pr.checkOrUpdatePodUID(podUID); err != nil {
-		return nil, err
+	pod, err := getPod(podLister, kclient, pr.PodNamespace, pr.PodName)
+	if err == nil && !IsStaticPod(pod) {
```
Comment (tssurya) [inline go-controller/pkg/cni/cni.go:140]: coming back to the main logic here, was the decision to totally skip the check for static pods or was it to use the mirror hash: kubernetes.io/config.mirror: a456b77dd3038db3f5590ca665f2e3b4 versus the uid in the cni request: 
```
"status": {
    "id": "196c503c05f592ac97a454732b2839e914eb1a85da6cc270009b31a72d008557",
    "metadata": {
      "attempt": 0,
      "name": "kube-apiserver-ovn-control-plane",
      "namespace": "kube-system",
      "uid": "a456b77dd3038db3f5590ca665f2e3b4"
    },
```
instead of the uid checks we do for normal pods.

```
hash := getPodHash(pod)
	copyPod.Annotations[kubetypes.ConfigMirrorAnnotationKey] = hash

func getPodHash(pod *v1.Pod) string {
	// The annotation exists for all static pods.
	return pod.Annotations[kubetypes.ConfigHashAnnotationKey]
}
```

**Review by tssurya — CHANGES_REQUESTED** [review]

File: go-controller/pkg/cni/cni.go (hunk 1 above)
Comment (flavio-fernandes) [reply go-controller/pkg/cni/cni.go:141]: ah, ack.

File: go-controller/pkg/cni/utils.go [hunk 7]
```diff
@@ -131,3 +131,40 @@ func PodAnnotation2PodInfo(podAnnotation map[string]string, checkExtIDs bool, po
 	}
 	return podInterfaceInfo, nil
 }
+
+//START taken from https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/types/pod_update.go
+const (
```
[the remark this answers is not in the document]
Comment (bpickard22) [reply go-controller/pkg/cni/utils.go:136]: since this was taken straight from upstream kube, I didnt know if I should make changes?

File: go-controller/pkg/cni/cni.go (hunk 6 above)
Comment (bpickard22) [reply go-controller/pkg/cni/cni.go:140]: we need to skip the check for static pods, as well as use the uid that master is expecting from the node side

**Comment by bpickard22** [comment]
leaving this as wip until i add tests, but I was able to successfully create a static pod in kind with these changes

File: go-controller/pkg/cni/helper_linux.go [hunk 8]
```diff
@@ -324,6 +324,14 @@ func ConfigureOVS(ctx context.Context, namespace, podName, hostIfaceName string,
 
 	ifaceID := util.GetIfaceId(namespace, podName)
 	initialPodUID := ifInfo.PodUID
+	pod, err := getPod(podLister, kclient, namespace, podName)
+	if IsStaticPod(pod) {
+		if err == nil {
+			initialPodUID = string(pod.ObjectMeta.UID)
+		} else {
+			klog.Infof("Could not find pod %v %v", pod, err)
```
Comment (flavio-fernandes) [inline go-controller/pkg/cni/helper_linux.go:332]: not Info... maybe a Warning?
Interestingly, here we are not paying attention if getPod failed when checking IsStaticPod. Maybe do the same at the other place where this is being called? We should be consistent about how/when we handle failures in getPod, agree?

**Review by flavio-fernandes — CHANGES_REQUESTED** [review]
some nitpick. :)
Please add test and consider squashing the 2nd and 3rd commits into one, as they go together.

File: go-controller/pkg/cni/cni.go [hunk 9]
```diff
@@ -136,8 +136,11 @@ func (pr *PodRequest) cmdAdd(kubeAuth *KubeAPIAuth, podLister corev1listers.PodL
...
+		if err := pr.checkOrUpdatePodUID(podUID); err != nil {
```
[the remark this answers is not in the document]
Comment (bpickard22) [reply go-controller/pkg/cni/cni.go:141]: need the := for same reason as before

File: go-controller/pkg/cni/helper_linux.go (hunk 8 above)
Comment (bpickard22) [reply go-controller/pkg/cni/helper_linux.go:332]: Agree, changed to warning and followed same procedure on the other function call

File: go-controller/pkg/cni/utils.go [hunk 10]
```diff
@@ -134,21 +134,8 @@ func PodAnnotation2PodInfo(podAnnotation map[string]string, checkExtIDs bool, po
 
 //START taken from https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/types/pod_update.go
 const (
-	ConfigSourceAnnotationKey    = "kubernetes.io/config.source"
-	ConfigMirrorAnnotationKey    = kapi.MirrorPodAnnotationKey
-	ConfigFirstSeenAnnotationKey = "kubernetes.io/config.seen"
```
Comment (flavio-fernandes) [inline go-controller/pkg/cni/utils.go]: lets merge it to the first commit, so we never addit to begin with

Comment (bpickard22) [reply go-controller/pkg/cni/utils.go]: will do

File: go-controller/pkg/cni/utils.go [hunk 11]
```diff
@@ -134,21 +134,8 @@ func PodAnnotation2PodInfo(podAnnotation map[string]string, checkExtIDs bool, po
 
 //START taken from https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/types/pod_update.go
 const (
-	ConfigSourceAnnotationKey    = "kubernetes.io/config.source"
```
Comment (flavio-fernandes) [inline go-controller/pkg/cni/utils.go]: Let's make the changes to go-controller/pkg/cni/utils.go only be part of the first commit.

File: go-controller/pkg/cni/cni.go [hunk 12]
```diff
@@ -136,8 +136,13 @@ func (pr *PodRequest) cmdAdd(kubeAuth *KubeAPIAuth, podLister corev1listers.PodL
 	if err != nil {
 		return nil, fmt.Errorf("failed to get pod annotation: %v", err)
 	}
-	if err := pr.checkOrUpdatePodUID(podUID); err != nil {
-		return nil, err
+	pod, err := getPod(podLister, kclient, pr.PodNamespace, pr.PodName)
+	if err == nil && !IsStaticPod(pod) {
```
Comment (flavio-fernandes) [inline go-controller/pkg/cni/cni.go:140]: how about:
```
pod, err := getPod(podLister, kclient, pr.PodNamespace, pr.PodName)
	if !IsStaticPod(pod) {
		if err = pr.checkOrUpdatePodUID(podUID); err != nil {
			return nil, err
		}
	} else {
		klog.Warningf("Could not find pod %v %v", pod, err)
		return nil, err 
	}
```

**Comment by bpickard22** [comment]
@tssurya does running getPod twice here have a scale impact?

File: test/e2e/static_pods.go [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+	framework.Logf("Running command %v", cmd)
+	_, err := runCommand(cmd...)
+	if err != nil {
+		framework.Failf("failed to remove pod file from node %s", nodeName)
+	}
+
+}
+
+//This test does the following
+//Applies a staticPod.yaml file to a nodes /etc/kubernetes/manifest dir
```
Comment (flavio-fernandes) [inline test/e2e/static_pods.go:47]: `static-pod.yaml`

File: test/e2e/static_pods.go [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+		framework.Logf("creating %s on node %s", podName, nodeName)
+		var staticPodYaml = fmt.Sprintf(`apiVersion: v1
+kind: Pod
+metadata:
+  name: static-pod
+  namespace: %s
+spec: 
+  containers: 
+    - name: web 
+      image: centos:8
```
Comment (flavio-fernandes) [inline test/e2e/static_pods.go:84]: Use same image as other tests, so there is no need for cluster to kind load a new image: `k8s.gcr.io/e2e-test-images/agnhost:2.26`

**Review by flavio-fernandes — CHANGES_REQUESTED** [review]
nice!
some small final nits from me.
Also, can you please try this e2e on a kind cluster that does not have your other commits and make sure it fails?

File: test/e2e/static_pods.go [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+		createStaticPod(f, nodeName, staticPodYaml)
+		defer removeStaticPodFile(nodeName, podFile)
+		err = e2epod.WaitForPodSuccessInNamespace(f.ClientSet, podName, f.Namespace.Name)
+		if err != nil {
+			logs, logErr := e2epod.GetPodLogs(f.ClientSet, f.Namespace.Name, podName, contName)
+			if logErr != nil {
+				framework.Logf("Warning: Failed to get logs from pod %q: %v", podName, logErr)
+			} else {
+				framework.Logf("pod %s/%s logs:\n%s", f.Namespace.Name, podName, logs)
+			}
```
Comment (flavio-fernandes) [inline test/e2e/static_pods.go:96]: @bpickard22 I had an idea: can you make the removal of the pod part of the test itself, instead of the cleanup? That way we will test/verify that the static pood removal works as expected (CNI DEL codepath).

**Review by flavio-fernandes — COMMENTED** [review]
one more (final) ask from me. Sorry for not thinking of that sooner.
Then, we may be ready to remove the wip from the pr.
@tssurya PTAL

File: test/e2e/static_pods.go [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+	if err := ioutil.WriteFile(podFile, []byte(podYaml), 0644); err != nil {
+		framework.Failf("Unable to write static-pod.yaml  to disk: %v", err)
+	}
+	defer func() {
+		if err := os.Remove(podFile); err != nil {
+			framework.Logf("Unable to remove the static-pod.yaml from disk: %v", err)
+		}
+	}()
+	var dst = fmt.Sprintf("%s:/etc/kubernetes/manifests/%s", nodeName, podFile)
+	cmd := []string{"docker", "cp", podFile, dst}
```
Comment (dougbtv) [inline test/e2e/static_pods.go:27]: This is.... a possible edge case, that I'd consider optional to handle, but...

In the case that the kubelet starts creating the pod before all the bytes are written to the file -- it's possible it'll kick it off with only part of the file, and the pod will fail to come up, or come up in a failed state. You can avoid this by writing the file to a temporary location, e.g. `/etc/my.temp.manifests/static-pod.yml` and the moving it into the final destination a la `mv /etc/my.temp.manifests/static-pod.yml /etc/kubernetes/manifests/static-pod.yml` -- then it's an atomic operation.

IMO, I'd give it a half dozen rounds of testing and see if you hit this situation. But, you could avoid it entirely by doing the move for the atomic operation.

**Comment by dougbtv** [comment]
Also, overall approach looking good Ben :+1:

**Comment by bpickard22** [comment]
> nice! some small final nits from me. Also, can you please try this e2e on a kind cluster that does not have your other commits and make sure it fails?

failed on cluster without my changes @flavio-fernandes

File: test/e2e/static_pods.go (hunk 16 above)
Comment (dougbtv) [reply test/e2e/static_pods.go:27]: Ben and I chatted it out, and Ben didn't hit it after a number of tests. So, I think we're OK with the atomic operation. Thanks Ben!

Comment (flavio-fernandes) [reply test/e2e/static_pods.go:27]: Valid point @dougbtv and I agree we may not need to worry about it here. The reason being is that the pod creation does not really happen while file is written, but only after the `docker cp` command takes place.

File: test/e2e/static_pods.go [hunk 17]
```diff
@@ -0,0 +1,103 @@
...
+			logs, logErr := e2epod.GetPodLogs(f.ClientSet, f.Namespace.Name, podName, contName)
+			if logErr != nil {
+				framework.Logf("Warning: Failed to get logs from pod %q: %v", podName, logErr)
+			} else {
+				framework.Logf("pod %s/%s logs:\n%s", f.Namespace.Name, podName, logs)
+			}
+		}
+		ginkgo.By("Removing the pod file from the nodes /etc/kubernetes/manifests")
+		framework.Logf("Removing %s from %s", podName, nodeName)
+		removeStaticPodFile(nodeName, podFile)
```
Comment (flavio-fernandes) [inline test/e2e/static_pods.go:101]: cool beans. And now do a getPods to make sure it is really gone ... without any bangs! :)

**Review by flavio-fernandes — CHANGES_REQUESTED** [review]
last request from me. I promise! :)

File: test/e2e/static_pods.go (hunk 17 above)
Comment (bpickard22) [reply test/e2e/static_pods.go:100]: had to resolve this by checking for pod success in namespace and expecting an error. Couldnt use any of the expectPodTerminated or other deletion failures from the e2e podPackage because static pod deletion does not work the same as api pod deletion. When we remove the file the pod just ceases to exist. Not quite the same as deletion. The delete functions would all error out because its like the pod never existed in the first place, but we can check and see if that podname is running in the namespace, and if it timesout (meaning it is not) then the pod must have been removed :)

Comment (flavio-fernandes) [reply test/e2e/static_pods.go:100]: 👍

**Review by flavio-fernandes — COMMENTED** [review]
/lgtm
TY for your patience in adressing all my asks @bpickard22 ! Super happy to have this fix and the test in the codebase now.

File: go-controller/pkg/node/node_dpu_test.go [hunk 18]
```diff
@@ -164,7 +164,7 @@ var _ = Describe("Node DPU tests", func() {
 		It("Fails if dpu.connection-details Pod annotation is not present", func() {
 			pod.Annotations = map[string]string{}
 			fakeClient := newFakeKubeClientWithPod(&pod)
-			podNamespaceLister.On("Get", mock.AnythingOfType("string")).Return(pod, nil)
```
Comment (flavio-fernandes) [inline go-controller/pkg/node/node_dpu_test.go:167]: @dcbw based on commit d39716645fe8bc7f021e7ea0b5f61bbcf421fe8a ... can you please confirm this change is okay by you?

**Review by flavio-fernandes — COMMENTED** [review]
/remove-lgtm

**Review by jcaamano — COMMENTED** [review]
Sorry, late to the party and I might be adding more noise than anything given the number of reviews already in place. Commenting just in case.

File: go-controller/pkg/cni/cni.go [hunk 19]
```diff
@@ -142,14 +144,23 @@ func (pr *PodRequest) cmdAdd(kubeAuth *KubeAPIAuth, clientset *ClientSet, useOVS
 	}
 	// Get the IP address and MAC address of the pod
 	// for DPU, ensure connection-details is present
+<<<<<<< HEAD
+	pod, annotations, podNADAnnotation, err := getPodWithAnnotations(pr.ctx, podLister, kclient, namespace, podName,
+=======
 	podUID, annotations, podNADAnnotation, err := GetPodAnnotations(pr.ctx, clientset, namespace, podName,
+>>>>>>> ovn-org/master
```
Comment (flavio-fernandes) [inline go-controller/pkg/cni/cni.go:151]: !!

Comment (bpickard22) [reply go-controller/pkg/cni/cni.go]: @flavio-fernandes missed this conflict when merging master in, took care of it in a last commit. Currently cleaning up commit history

File: go-controller/pkg/cni/helper_linux.go [hunk 20]
```diff
@@ -354,6 +354,12 @@ func ConfigureOVS(ctx context.Context, namespace, podName, hostIfaceName string,
 		ifaceID = util.GetSecondaryNetworkIfaceId(namespace, podName, ifInfo.NADName)
 	}
 	initialPodUID := ifInfo.PodUID
+	pod, err := PodInfoGetter.getPod(getter, namespace, podName)
+	if err != nil {
+		klog.Warningf("Could not find pod %v %v", pod, err)
+	} else if IsStaticPod(pod) {
+		initialPodUID = string(pod.ObjectMeta.UID)
+	}
```
Comment (jcaamano) [inline go-controller/pkg/cni/helper_linux.go:362]: I don't think we need this change. 

`ifInfo.PodUID` should already be set to the correct pod UID:
In `checkOrUpdatePodUID` we set the pod UID in the PorRequest, then in  PodAnnotation2PodInfo is set to PodInterfaceInfo which we are using here.

File: go-controller/pkg/cni/cni.go [hunk 21]
```diff
@@ -80,11 +82,11 @@ func (pr *PodRequest) String() string {
...
 // But if the existing UID is valid and does not match the given UID then the
 // sandbox request is for a different pod instance and should be terminated.
-func (pr *PodRequest) checkOrUpdatePodUID(podUID string) error {
-	if pr.PodUID == "" {
-		// Runtime didn't pass UID, use the one we got from the pod object
-		pr.PodUID = podUID
-	} else if podUID != pr.PodUID {
+func (pr *PodRequest) checkOrUpdatePodUID(pod *kapi.Pod) error {
+	if pr.PodUID == "" || IsStaticPod(pod) {
+		// Runtime didn't pass UID, or the pod is a static pod, use the one we got from the pod object
```
Comment (jcaamano) [inline go-controller/pkg/cni/cni.go:87]: nit: Would it be god to leave a comment here for posterity?
```
// Static pod UID is a hash of the pod itself that does not match
// the UID of the mirror kubelet creates on the api /server. 
// We will use the UID of the mirror.
// The hash is annotated in the mirror pod (kubernetes.io/config.hash)
// and we could match against it, but let's avoid that for now as it is not 
// a published standard.
```

**Comment by jcaamano** [comment]
Sorry it got me some time to get back to this @bpickard22

**Comment by bpickard22** [comment]
> Sorry it got me some time to get back to this @bpickard22

nw @jcaamano thanks for looking

File: go-controller/pkg/cni/helper_linux.go (hunk 20 above)
Comment (bpickard22) [reply go-controller/pkg/cni/helper_linux.go:362]: we hit checkOrUpdatePodUID before we set the correct UID. When I remove the static pod check from the function, the static pod will not create with the pd deleted before sandbox op began error, so at this point the uid has not been updated.

File: go-controller/pkg/cni/cni.go (hunk 21 above)
Comment (bpickard22) [reply go-controller/pkg/cni/cni.go:93]: yes i think this makes sense

File: go-controller/pkg/cni/helper_linux.go (hunk 20 above)
Comment (bpickard22) [reply go-controller/pkg/cni/helper_linux.go]: you are right, removed the change

File: test/e2e/static_pods.go (hunk 17 above)
Comment (jcaamano) [reply test/e2e/static_pods.go:100]: @bpickard22 why can't we use  `[WaitForPodNotFoundInNamespace](https://pkg.go.dev/k8s.io/kubernetes/test/e2e/framework/pod#WaitForPodNotFoundInNamespace)`?: 
```
WaitForPodNotFoundInNamespace returns an error if it takes too long for the pod to fully terminate. Unlike `waitForPodTerminatedInNamespace`, the pod's Phase and Reason are ignored. If the pod Get api returns IsNotFound then the wait stops and nil is returned. If the Get api returns an error other than "not found" then that error is returned and the wait stops.
```

Trying to avoid the chance of a flake if kubelet does not react immediately after the manifest is removed from disk.

Comment (bpickard22) [reply test/e2e/static_pods.go:109]: Note: going to change this to use the provided function from the framework package in a seperate pr, where i bump the versions of the deps in the go mod to 1.26

**Review by jcaamano — APPROVED** [review]

**Comment by flavio-fernandes** [comment]
🥳
